[docs][AMDGPU] Document amdgpu.buffer.oob.mode more, especially on fat pointers (#134734)
llvm/llvm-project: llvm/docs/AMDGPUUsage.rst

                             other operations within the same address space.
     ======================= ===================================================

Relaxed Buffer OOB (Out Of Bounds) Mode
---------------------------------------

Instructions that load from or store to buffer resources (and thus, by extension
buffer fat pointers and buffer strided pointers) generally implement handling for
out of bounds (OOB) memory accesses, including those that are partially OOB,
if the buffer resource resource has the required flags set.

When operating on more than 32 bits of data, the `voffset` used for the access
will be range-checked for each 32-bit word independently. This check uses saturating
arithmetic and interprets the offset as an unsigned value.

The behavior described above conflicts with the ABI requirements of certain graphics
APIs that require out of bounds accesses to be handled strictly so that accessed

> that begin out of bounds but then access in-bounds elements (such as loading A

Reviewer: ```suggestion
that begin out of bounds but then access in-bounds elements (such as loading a
```

> ``<4 x i32>`` beginning at offset ``-4``) still load the three in-bounds integers.

Reviewer: Just out of curiosity, will it be `<undef, i32, i32, 32>` or `<i32, i32, i32, undef>`?

Author: `<0, i32, i32, i32>` if you're using whatever the strict Vulkan API is

Ordinary, the left end gives you `<0, 0, 0, 0,>`

Author: Added to docs

Similarly, buffer fat pointers permit operating types such as `<8 x i8>` which
must be accessed (and bounds-checked) 4 bytes at a time. Non-word-aligned
accesses to such types from near the end of a buffer resource (such as starting

> a load of an ``<8xi8>`` from an offset of ``6`` on an 8-byte buffer) will treat

Reviewer: ```suggestion
a load of an ``<8 x i8>`` from an offset of ``6`` on an 8-byte buffer) will treat
```

the initial two bytes to be loaded/stored as out of bounds, even though, under
a strict interpretation of the bounds-checking semantics, they would be out of bounds.

These violations of strict bounds-checking semantics for buffer resources require

> usage of less-vectorized code to ensure correctness. Ifthis strict conformance

Reviewer: ```suggestion
usage of less-vectorized code to ensure correctness. If this strict conformance
```

is not required, the target feature ``relaxed-oob-buffer-mode`` should be enabled
(using ``-mcpu``, ``-offload-arch`` or ``-mattr``).

``relaxed-buffer-oob-mode`` permits unaligned memory acceses through a buffer resource

> to propagate to nearby elemennts, causing them to become out of bounds as well.

Reviewer 2: Typo
```suggestion
to propagate to nearby elements, causing them to become out of bounds as well.
```

                            other operations within the same address space.
    ======================= ===================================================

.. _amdgpu-fat-buffer-oob-handling:

Buffer Fat Pointer out of bounds (OOB) handling
-----------------------------------------------

Instructions that load from or store to buffer resources (and thus, by extension
buffer fat pointers and buffer strided pointers) generally implement handling for
out of bounds (OOB) memory accesses, including those that are partially OOB,
if the buffer resource resource has the required flags set. How ordinary
``load`` and ``store`` instructions are lowered to buffer operations is partly
controlled by the ``amdgpu.buffer.oob.mode`` (see :ref:`amdgpu-module-flags`). If
that flag is set to ``1`` (relaxed), no handling to improve the expected behavior
of OOB accesses is performed, while if it is set to ``0`` (any) or ``2`` (strict),

> oerations may be spliit to ensure correctness under stronger models of how

Reviewer 3: ```suggestion
oerations may be split to ensure correctness under stronger models of how
```

out-of-bounds accesses should behave.

When operating on more than 32 bits of data, the ``voffset`` used for the access
will be range-checked for each 32-bit word independently. This check uses saturating
arithmetic and interprets the offset as an unsigned value.

The behavior described above conflicts with the ABI requirements of certain graphics
APIs that require out of bounds accesses to be handled strictly so that accessed
that begin out of bounds but then access in-bounds elements (such as loading a
``<4 x i32>`` beginning at offset ``-4``) still load the three in-bounds integers
(producing ``<0, v0, v1, v2>`` and not ``<0, 0, 0, 0>``. So, under strict OOB
handling, such an access will be split into four ``i32`` accesses. Note this this
can only happen for underaligned loads - such wraparound isn't possible for

> loads thatare alligned to their natural size.

Reviewer 3: ```suggestion
loads that are aligned to their natural size.
```

> Similarly, buffer fat pointers permit operating types such as ``<8 x i8>`` which

Reviewer 3: Missing a word "operating types"

must be accessed (and bounds-checked) 4 bytes at a time. Non-word-aligned
accesses to such types from near the end of a buffer resource (such as starting
a load of an ``<8 x i8>`` from an offset of ``6`` on an 8-byte buffer) will treat
the initial two bytes to be loaded/stored as out of bounds, even though, under
a strict interpretation of the bounds-checking semantics, they would be in bounds.
Under strict OOB handling, such a load will be split into a sequence of ``<2 x i16>``
loads.

.. note::

  These transformations do not impact buffer intrinsic calls (calling

>   ``llvm.amdgcn.raw.buffer.*`` directly).

Reviewer 3: Note is a bit confusing, this changes what transformations could occur to the intrinsics

Author: Yeah, I think there's other wording here. 

What I want to say, is "Buffer OOB mode will not protect you from malformed intrinsic calls. For example, trying to do an underaligned vector load with `raw[.ptr].buffer.load` will lead to it being emitted even if you get unexpected results. We will lower to the correct thing when dealing with `ptr(7)/ptr(9)` though.".

Author: Found the better wording
